### ISSUE TITLE
Track and Troubleshoot Datadog Database APM Support Ticket

### ISSUE THREAD

**Issue by olivereri** [department-of-veterans-affairs/va.gov-cms#8865]
## Description
Datadog APM for Databases isn't working after configuration on CMS-Test Dev. Opened a support ticket. Will create a separate ticket to track and implement solutions
https://help.datadoghq.com/hc/en-us/requests/795564

## Acceptance Criteria
- [x] Datadog Support has been engaged on the issue.
- [x] Issue has been resolved with Datadog Database APM configured on all envs (dev,staging,prod) and reporting metrics to the DD Dashboard.

**Comment by olivereri** [comment]
- Reconfigured CMS-test Dev datadog-dagent to do database performance monitoring.
- Reconfigured CMS-test Dev Database parameter groups match MySQL 5.6/5.7 settings.
- Verified database database is configured per documentation.
- Re-opened Datadog ticket and sent log/config settings to datadog using the `flare` command https://docs.datadoghq.com/agent/troubleshooting/send_a_flare/?tab=agentv6v7

**Comment by olivereri** [comment]
Amanda White
Yesterday at 16:15

Hi Eric,
 
Thanks for following up on this issue. My name is Amanda and I'll be handling this follow-up.
 
After taking a look at the flare you previously sent over, I'm seeing 2 key Warnings in the mysql Check that is resulting in no data being shown:
 
mysql (8.1.0)
    -------------
      Instance ID: mysql:c513073659b31563 [WARNING]
      Configuration Source: file:/etc/datadog-agent/conf.d/mysql.d/conf.yaml
      Total Runs: 107
      Metric Samples: Last Run: 114, Total: 12,182
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 3, Total: 321
      Average Execution Time : 38ms
      Last Execution Date : 2022-05-10 20:59:26 UTC (1652216366000)
      Last Successful Execution Date : 2022-05-10 20:59:26 UTC (1652216366000)
      metadata:
        flavor: MariaDB
        version.build: log
        version.major: 10
        version.minor: 5
        version.patch: 13
        version.raw: 10.5.13+log
        version.scheme: semver
      Warning: Privileges error getting replication status (must grant REPLICATION CLIENT): (1227, 'Access denied; you need (at least one of) the SUPER, SLAVE MONITOR privilege(s) for this operation')
      Warning: Unable to collect statement metrics because the performance schema is disabled. See https://docs.datadoghq.com/database_monitoring/setup_mysql/troubleshooting#performance-schema-not-enabled for more details
code=performance-schema-not-enabled host=dsva-cms-test-dev-db.cdqbofmbcmtd.us-gov-west-1.rds.amazonaws.com
 
Warning 1: Privileges error getting replication status (must grant REPLICATION CLIENT): (1227, 'Access denied; you need (at least one of) the SUPER, SLAVE MONITOR privilege(s) for this operation')
 
To action this issue - you will need to log into the SQL server with an admin user.

1. Verify the datadog user:
mysql -u datadog --password=<UNIQUEPASSWORD> -e "show slave status" && \
echo -e "\033[0;32mMySQL grant - OK\033[0m" || \
echo -e "\033[0;31mMissing REPLICATION CLIENT grant\033[0m"
2. Grant permissions to the user with the following commands:
mysql> GRANT REPLICATION CLIENT ON *.* TO 'datadog'@'localhost' WITH MAX_USER_CONNECTIONS 5;
Query OK, 0 rows affected, 1 warning (0.00 sec)
mysql> GRANT PROCESS ON *.* TO 'datadog'@'localhost';
Query OK, 0 rows affected (0.00 sec) 
Please be sure to enter these commands with an admin user on your sql server. For more information on this you can have a look at our public documentation. MySQL
 
Warning 2: Unable to collect statement metrics because the performance schema is disabled. See https://docs.datadoghq.com/database_monitoring/setup_mysql/troubleshooting#performance-schema-not-enabled for more details
code=performance-schema-not-enabled host=dsva-cms-test-dev-db.cdqbofmbcmtd.us-gov-west-1.rds.amazonaws.com
 
Per our Troubleshooting doc, the Agent requires the performance_schema option to be enabled. It is enabled by default by MySQL, but may be disabled in configuration or by your cloud provider. Follow the setup instructions for enabling it.
 
Please let me know that if after making the appropriate changes if you're able to see data populating. If not, I'm more than happy to continue troubleshooting this issue for you!
 
Feel free to reach out with any questions.
 
Best,
Amanda White | Solutions Engineer | Datadog

Datadog Support Platform
Datadog Documentation

**Comment by olivereri** [comment]
Amanda White
Yesterday at 16:07
Hi Eric,
 
Thanks for getting back to me! I have noted that you're wanting to get this working not using pure MySQL. Since MariaDB + AWS RDS is not officially supported, this is what's causing the configuration to be a bit tricky. That being said, I'm confident we can get this working for you!
 
I do have one quick question. What version of MariaDB is being used? With other customer's that have attempted a similar setup, the MariaDB version compatibility can affect what you're trying to accomplish.
 
As soon as you get back to me with the version, I'm going to consult with our internal teams to continue to troubleshoot this issue. In the meantime if you have any other questions please let me know.
 
Best,
Amanda White | Solutions Engineer | Datadog

======================================================

Eric Oliver
Yesterday at 20:04
That's wonderful news! The Database Engine version is 10.5.13

**Comment by olivereri** [comment]
Amanda White
Today at 09:27

Hi Eric,
 
I've taken this issue to our internal teams for some additional investigation and troubleshooting. I'll reach out as soon as I have an update!
 
Best,
Amanda White | Solutions Engineer | Datadog

Datadog Support Platform
Datadog Documentation

**Comment by olivereri** [comment]
Eric Oliver
Today at 11:23

Hi Amanda,

Like I mentioned my SQL-fu isn't strong so I wasn't sure how to query the performance_schema DB Table events_statements_summary_by_digest to verify that it's actually on. Instead I've probably over shared info that might be unneeded. I've attached the the entire output of that table. I've also attached a screenshot from the AWS RDS Dashboard for the param group. Additionally, I found some MariaDB documentation on verifying if performance_schema is enabled:

MariaDB [(none)]> use performance_schema;
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Database changed
MariaDB [performance_schema]> SHOW VARIABLES LIKE 'performance_schema';
+--------------------+-------+
| Variable_name | Value |
+--------------------+-------+
| performance_schema | ON |
+--------------------+-------+
1 row in set (0.01 sec)

MariaDB [performance_schema]> SHOW VARIABLES LIKE "perf%";
+----------------------------------------------------------+-------+
| Variable_name | Value |
+----------------------------------------------------------+-------+
| performance_schema | ON |
| performance_schema_accounts_size | -1 |
| performance_schema_digests_size | 10000 |
| performance_schema_events_stages_history_long_size | 10000 |
| performance_schema_events_stages_history_size | 10 |
| performance_schema_events_statements_history_long_size | 10000 |
| performance_schema_events_statements_history_size | 10 |
| performance_schema_events_transactions_history_long_size | 10000 |
| performance_schema_events_transactions_history_size | 10 |
| performance_schema_events_waits_history_long_size | 10000 |
| performance_schema_events_waits_history_size | 10 |
| performance_schema_hosts_size | -1 |
| performance_schema_max_cond_classes | 90 |
| performance_schema_max_cond_instances | -1 |
| performance_schema_max_digest_length | 4096 |
| performance_schema_max_file_classes | 80 |
| performance_schema_max_file_handles | 32768 |
| performance_schema_max_file_instances | -1 |
| performance_schema_max_index_stat | -1 |
| performance_schema_max_memory_classes | 320 |
| performance_schema_max_metadata_locks | -1 |
| performance_schema_max_mutex_classes | 210 |
| performance_schema_max_mutex_instances | -1 |
| performance_schema_max_prepared_statements_instances | -1 |
| performance_schema_max_program_instances | -1 |
| performance_schema_max_rwlock_classes | 50 |
| performance_schema_max_rwlock_instances | -1 |
| performance_schema_max_socket_classes | 10 |
| performance_schema_max_socket_instances | -1 |
| performance_schema_max_sql_text_length | 4096 |
| performance_schema_max_stage_classes | 160 |
| performance_schema_max_statement_classes | 222 |
| performance_schema_max_statement_stack | 10 |
| performance_schema_max_table_handles | -1 |
| performance_schema_max_table_instances | -1 |
| performance_schema_max_table_lock_stat | -1 |
| performance_schema_max_thread_classes | 50 |
| performance_schema_max_thread_instances | -1 |
| performance_schema_session_connect_attrs_size | 512 |
| performance_schema_setup_actors_size | -1 |
| performance_schema_setup_objects_size | -1 |
| performance_schema_users_size | -1 |
+----------------------------------------------------------+-------+
42 rows in set (0.00 sec)

**Comment by ENeal49** [comment]
blocked due to waiting on Datadog

**Comment by olivereri** [comment]
Amanda White
Today at 09:42

Hi Eric,
 
Thanks for sending all of that over. After taking a look at it all, I've decided to escalate this issue to our product and engineering teams. I'll reach out again as soon as I have an update or if we need more information.
 
Thanks for your patience!
 
Best,
Amanda White | Solutions Engineer | Datadog

**Comment by olivereri** [comment]
Amanda White
Today at 09:39
Hi Eric,
 
I've heard back from our product team, and we'll need a bit more info:
 
Can you run SHOW VARIABLES LIKE 'performance_schema' directly on the database?

 
Best,
Amanda White | Solutions Engineer | Datadog

**Comment by olivereri** [comment]
Eric Oliver
Today at 12:27

Which database? If I don't use a specific database this is the output:

MariaDB [(none)]> SHOW VARIABLES LIKE 'performance_schema';
+--------------------+-------+
| Variable_name | Value |
+--------------------+-------+
| performance_schema | ON |
+--------------------+-------+
1 row in set (0.00 sec)

Below are the available databases. The output I show in my last post used performance_schema

MariaDB [(none)]> show databases;
+--------------------+
| Database |
+--------------------+
| datadog |
| dsva_cms_dev |
| information_schema |
| innodb |
| mysql |
| performance_schema |
| tmp |
+--------------------+
7 rows in set (0.01 sec)

Here's the query output from each of them:

MariaDB [(none)]> use datadog;
Database changed
MariaDB [datadog]> SHOW VARIABLES LIKE 'performance_schema';
+--------------------+-------+
| Variable_name | Value |
+--------------------+-------+
| performance_schema | ON |
+--------------------+-------+
1 row in set (0.00 sec)

MariaDB [datadog]> use dsva_cms_dev;
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Database changed
MariaDB [dsva_cms_dev]> SHOW VARIABLES LIKE 'performance_schema';
+--------------------+-------+
| Variable_name | Value |
+--------------------+-------+
| performance_schema | ON |
+--------------------+-------+
1 row in set (0.00 sec)

MariaDB [dsva_cms_dev]> use information_schema;
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Database changed
MariaDB [information_schema]> SHOW VARIABLES LIKE 'performance_schema';
+--------------------+-------+
| Variable_name | Value |
+--------------------+-------+
| performance_schema | ON |
+--------------------+-------+
1 row in set (0.00 sec)

MariaDB [information_schema]> use innodb;
Database changed
MariaDB [innodb]> SHOW VARIABLES LIKE 'performance_schema';
+--------------------+-------+
| Variable_name | Value |
+--------------------+-------+
| performance_schema | ON |
+--------------------+-------+
1 row in set (0.00 sec)

MariaDB [innodb]> use mysql;
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Database changed
MariaDB [mysql]> SHOW VARIABLES LIKE 'performance_schema';
+--------------------+-------+
| Variable_name | Value |
+--------------------+-------+
| performance_schema | ON |
+--------------------+-------+
1 row in set (0.00 sec)

MariaDB [mysql]> use performance_schema;
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Database changed
MariaDB [performance_schema]> SHOW VARIABLES LIKE 'performance_schema';
+--------------------+-------+
| Variable_name | Value |
+--------------------+-------+
| performance_schema | ON |
+--------------------+-------+
1 row in set (0.00 sec)

MariaDB [performance_schema]> use tmp;
Database changed
MariaDB [tmp]> SHOW VARIABLES LIKE 'performance_schema';
+--------------------+-------+
| Variable_name | Value |
+--------------------+-------+
| performance_schema | ON |
+--------------------+-------+
1 row in set (0.00 sec)

Is this what the team is looking for?

**Comment by olivereri** [comment]
Amanda White
Today at 12:48

Hi Eric,
 
Let me send this over to them and have them confirm!
 
Best,
Amanda White | Solutions Engineer | Datadog

Datadog Support Platform
Datadog Documentation

Avatar	
Amanda WhiteToday at 14:47

==================================================================

Hi Eric,
 
I've heard back and can confirm you did the right thing and those results look good! Based on this I do have a quick question, has the Agent been restarted since making any changes? Either way, would you mind restarting the Agent?
 
Let me know if you have any questions, and thanks again for your patience!
 
Best,
Amanda White | Solutions Engineer | Datadog

Datadog Support Platform
Datadog Documentation
Avatar	

==================================================================

Eric Oliver
Today at 15:05
Logs attached from i-0247c7fc812aa38d6

 datadog-agent-2022-05-19T22-05-34Z.zip (1000 KB)
Avatar	

==================================================================

Eric Oliver
Today at 15:06

Ok, I've restarted the Agent and sent another flare.

**Comment by olivereri** [comment]
Amanda White
Today at 15:12

Good news!! It looks like we've got it setup properly! (link)
 

https://a.cl.ly/04uQbBQ9
 
That being said it will take a bit for data to begin populating, so let's give it a bit and hopefully that did the trick!
 
Best,
Amanda White | Solutions Engineer | Datadog

Datadog Support Platform
Datadog Documentation
